### ISSUE TITLE
Add UserLoggedIn hook to SignInManager

### ISSUE THREAD

**Issue by rhegner** [aspnet/Identity#646]
I'm writing my custom SignInManager which is derived from `Microsoft.AspNet.Identity.SignInManager`. Right now I'm looking for a way to perform some action whenever the user has successfully logged in, regardless of the method he used (password, external, with/without two factor). That action could for example be setting a `LastLogin` field in my `ApplicationUser`.

I could override `SignInAsync` but it looks like this method not only gets called when the user has explicitly logged in but also on other occasions.

Or I could override `PasswordSignInAsync`, `TwoFactorSignInAsync` and `ExternalLoginSignInAsync`. Here I have to look at the sources to find out which overload of `PasswordSignInAsync` is the right one to override (or I just override all of its overloads), but this is not so future proof, since these implementation details could change in future versions.

Long story short: I think it would be nice to have some hook in `SignInManager` which gets called whenever a user has logged in.

**Comment by rustd** [comment]
Would #650 meet your requirements?

**Comment by rhegner** [comment]
No because I also need to perform some project specific actions after login.

**Comment by HaoK** [comment]
As discussed with @rustd last week

**Comment by HaoK** [comment]
Unassigned since this is in backlog, reassign when we want to do this.

**Comment by musabsudan** [comment]
adding that would be great, I noticed that it's a common pattern that leads people o custom implementations, some "hooks" to perform extra work would be fantastic.

Thanks

**Comment by HaoK** [comment]
Given that the current virtual SignInAsync catches too many flows for you, I think the best approach would be for you to extend exactly which methods you want to do your additional post SignIn logic.

**Comment by HaoK** [comment]
All the hooks should already be there as every method is already virtual
